simulators/eth2/engine/proxies.go

```diff
@@ -0,0 +1,77 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/rauljordan/engine-proxy/proxy"
+)
+
+type Proxy struct {
+	IP        net.IP
+	port      int
+	proxy     *proxy.Proxy
+	JWTSecret string
+	config    *proxy.SpoofingConfig
+
+	rpc *rpc.Client
+	mu  sync.Mutex
+}
+
+func NewProxy(hostIP net.IP, port int, destination string, jwtSecret []byte) *Proxy {
+	host := "0.0.0.0"
+	config := proxy.SpoofingConfig{}
+	options := []proxy.Option{
+		proxy.WithHost(host),
+		proxy.WithPort(port),
+		proxy.WithDestinationAddress(destination),
+		proxy.WithSpoofingConfig(&config),
+		proxy.WithJWTSecret(jwtSecret),
+	}
+	proxy, err := proxy.New(options...)
+	if err != nil {
+		panic(err)
+	}
+	go func() {
+		if err := proxy.Start(context.Background()); err != nil {
+			panic(err)
+		}
+	}()
+	time.Sleep(100 * time.Millisecond)
+	log.Info("Starting new proxy", "host", host, "port", port)
+	return &Proxy{IP: hostIP, port: port, proxy: proxy, config: &config}
+}
+
+func (p *Proxy) UserRPCAddress() (string, error) {
+	return fmt.Sprintf("http://%v:%d", p.IP, p.port), nil
+}
+
+func (p *Proxy) EngineRPCAddress() (string, error) {
+	return fmt.Sprintf("http://%v:%d", p.IP, p.port), nil
+}
+
+func (p *Proxy) AddRequest(spoof *proxy.Spoof) {
```

Comment:
I think it would be great to have the option to add a callback, something like `func (p *Proxy) AddRequestCallback(callback func(interface{}) *proxy.Spoof)` , and the callback function receives the request and you can respond on a request-by-request basis whether to spoof or not.

I think at the moment `AddRequests` spoofs every single request it receives from the CL which is ok, but being able to decide on which requests to spoof would let us fine-tune the test cases. E.g. a test case where only the transition payload is corrupted, but every payload after that is ok.

Reply:
Yes, I thought about this too, but unfortunately raul's proxy does not support that. I can try to implement it in the proxy so we can use it in hive eventually

Reply:
I've added a PR to raul's repo to add this functionality: https://github.com/rauljordan/engine-proxy/pull/4
I haven't tested the changes but will do tomorrow.